File: DriftGame/uracer-desktop/src/com/bitfire/postprocessing/effects/Bloom.java
package com.bitfire.postprocessing.effects;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.glutils.FrameBuffer;
import com.bitfire.postprocessing.PostProcessor;
import com.bitfire.postprocessing.PostProcessorEffect;
import com.bitfire.postprocessing.filters.Blur;
import com.bitfire.postprocessing.filters.Blur.BlurType;
import com.bitfire.postprocessing.filters.Combine;
import com.bitfire.postprocessing.filters.Threshold;
import com.bitfire.postprocessing.utils.PingPongBuffer;

public final class Bloom extends PostProcessorEffect {
    private final PingPongBuffer pingPongBuffer;
    private final Blur blur;
    private final Threshold threshold;
    private final Combine combine;

    public Bloom(int fboWidth, int fboHeight) {
        pingPongBuffer = PostProcessor.newPingPongBuffer(fboWidth, fboHeight, PostProcessor.getFramebufferFormat(), false);

        blur = new Blur(fboWidth, fboHeight);
        threshold = new Threshold();
        combine = new Combine();

        setSettings(new Settings("default", 2, 0.277f, 1f, .85f, 1.1f, .85f));
    }

    @Override
    public void dispose() {
        combine.dispose();
        threshold.dispose();
        blur.dispose();
        pingPongBuffer.dispose();
    }

    public void setBaseIntensity(float intensity) {
        combine.setSource1Intensity(intensity);
    }

    public void setBloomIntensity(float intensity) {
        combine.setSource2Intensity(intensity);
    }

    public void setThreshold(float gamma) {
        threshold.setTreshold(gamma);
    }

    public void setBaseSaturation(float saturation) {
        combine.setSource1Saturation(saturation);
    }

    public void setBloomSaturation(float saturation) {
        combine.setSource2Saturation(saturation);
    }

    public void setBlurType(BlurType type) {
        blur.setType(type);
    }

    public void setSettings(Settings settings) {

        // setup threshold filter
        setThreshold(settings.bloomThreshold);

        // setup combine filter
        setBaseIntensity(settings.baseIntensity);
        setBaseSaturation(settings.baseSaturation);
        setBloomIntensity(settings.bloomIntensity);
        setBloomSaturation(settings.bloomSaturation);

        // setup blur filter
        setBlurPasses(settings.blurPasses);
        setBlurAmount(settings.blurAmount);
        setBlurType(settings.blurType);
    }

    public void setBlurPasses(int passes) {
        blur.setPasses(passes);
    }

    public void setBlurAmount(float amount) {
        blur.setAmount(amount);
    }

    @Override
    public void render(final FrameBuffer src, final FrameBuffer dest) {
        Texture texsrc = src.getColorBufferTexture();

        boolean blendingWasEnabled = PostProcessor.isStateEnabled(GL20.GL_BLEND);
        Gdx.gl.glDisable(GL20.GL_BLEND);

        pingPongBuffer.begin();
        {
            // threshold / high-pass filter
            // only areas with pixels >= threshold are blit to smaller fbo
            threshold.setInput(texsrc).setOutput(pingPongBuffer.getSourceBuffer()).render();

            // blur pass
            blur.render(pingPongBuffer);
        }
        pingPongBuffer.end();

        if (blendingWasEnabled) {
            Gdx.gl.glEnable(GL20.GL_BLEND);
        }

        restoreViewport(dest);

        // mix original scene and blurred threshold, modulate via
        // set(Base|Bloom)(Saturation|Intensity)
        combine.setOutput(dest).setInput(texsrc, pingPongBuffer.getResultTexture()).render();
    }

    @Override
    public void rebind() {
        blur.rebind();
        threshold.rebind();
        combine.rebind();
        pingPongBuffer.rebind();
    }

    public static class Settings {
        public final String name;

        public final BlurType blurType;
        public final int blurPasses; // simple blur
        public final float blurAmount; // normal blur (1 pass)
        public final float bloomThreshold;

        public final float bloomIntensity;
        public final float bloomSaturation;
        public final float baseIntensity;
        public final float baseSaturation;

        public Settings(String name, BlurType blurType, int blurPasses, float blurAmount, float bloomThreshold,
                        float baseIntensity, float baseSaturation, float bloomIntensity, float bloomSaturation) {
            this.name = name;
            this.blurType = blurType;
            this.blurPasses = blurPasses;
            this.blurAmount = blurAmount;

            this.bloomThreshold = bloomThreshold;
            this.baseIntensity = baseIntensity;
            this.baseSaturation = baseSaturation;
            this.bloomIntensity = bloomIntensity;
            this.bloomSaturation = bloomSaturation;
        }

        // simple blur
        public Settings(String name, int blurPasses, float bloomThreshold, float baseIntensity, float baseSaturation,
                        float bloomIntensity, float bloomSaturation) {
            this(name, BlurType.Gaussian5x5b, blurPasses, 0, bloomThreshold, baseIntensity, baseSaturation, bloomIntensity,
                    bloomSaturation);
        }
    }
}

File: DriftGame/uracer-desktop/src/com/bitfire/postprocessing/effects/Bloom.kt
package com.bitfire.postprocessing.effects

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.glutils.FrameBuffer
import com.bitfire.postprocessing.PostProcessor
import com.bitfire.postprocessing.PostProcessorEffect
import com.bitfire.postprocessing.filters.Blur
import com.bitfire.postprocessing.filters.Combine
import com.bitfire.postprocessing.filters.Threshold

class Bloom(fboWidth: Int, fboHeight: Int) : PostProcessorEffect() {

    private val pingPongBuffer = PostProcessor.newPingPongBuffer(fboWidth, fboHeight, PostProcessor.framebufferFormat, false)
    private val blur = Blur(fboWidth, fboHeight)
    private val threshold = Threshold()
    private val combine = Combine()

    init {
        setSettings(Settings("default", 2, 0.277F, 1F, .85F, 1.1F, .85F))
    }

    override fun dispose() {
        combine.dispose()
        threshold.dispose()
        blur.dispose()
        pingPongBuffer.dispose()
    }

    fun setBaseIntensity(intensity: Float) {
        combine.setSource1Intensity(intensity)
    }

    fun setBloomIntensity(intensity: Float) {
        combine.setSource2Intensity(intensity)
    }

    fun setThreshold(gamma: Float) {
        threshold.setTreshold(gamma)
    }

    fun setBaseSaturation(saturation: Float) {
        combine.setSource1Saturation(saturation)
    }

    fun setBloomSaturation(saturation: Float) {
        combine.setSource2Saturation(saturation)
    }

    fun setBlurType(type: Blur.BlurType) {
        blur.setType(type)
    }

    fun setSettings(settings: Settings) {

        // The threshold filter
        setThreshold(settings.bloomThreshold)

        // The combine filter
        setBaseIntensity(settings.baseIntensity)
        setBaseSaturation(settings.baseSaturation)
        setBloomIntensity(settings.bloomIntensity)
        setBloomSaturation(settings.bloomSaturation)

        // The blur filter
        setBlurPasses(settings.blurPasses)
        setBlurAmount(settings.blurAmount)
        setBlurType(settings.blurType)
    }

    fun setBlurPasses(passes: Int) {
        blur.passes = passes
    }

    fun setBlurAmount(amount: Float) {
        blur.setAmount(amount)
    }

    override fun render(src: FrameBuffer, dest: FrameBuffer?) {
        val texsrc = src.colorBufferTexture

        val blendingWasEnabled = PostProcessor.isStateEnabled(GL20.GL_BLEND)
        Gdx.gl.glDisable(GL20.GL_BLEND)

        pingPongBuffer.begin()

        // The threshold filter (Only areas with pixels >= threshold are blit to smaller fbo).
        threshold.setInput(texsrc).setOutput(pingPongBuffer.sourceBuffer).render()

        blur.render(pingPongBuffer)
        pingPongBuffer.end()

        if (blendingWasEnabled) Gdx.gl.glEnable(GL20.GL_BLEND)
        restoreViewport(dest)

        // mix original scene and blurred threshold, modulate via set(Base|Bloom)(Saturation|Intensity)
        combine.setOutput(dest).setInput(texsrc, pingPongBuffer.resultTexture).render()
    }

    override fun rebind() {
        blur.rebind()
        threshold.rebind()
        combine.rebind()
        pingPongBuffer.rebind()
    }

    class Settings(
        val name: String, val blurType: Blur.BlurType, val blurPasses: Int, val blurAmount: Float, val bloomThreshold: Float, val baseIntensity: Float, val baseSaturation: Float, val bloomIntensity: Float, val bloomSaturation: Float
    ) {
        constructor(name: String, blurPasses: Int, bloomThreshold: Float, baseIntensity: Float, baseSaturation: Float, bloomIntensity: Float, bloomSaturation: Float) : this(
            name, Blur.BlurType.Gaussian5x5b, blurPasses, 0f, bloomThreshold, baseIntensity, baseSaturation, bloomIntensity, bloomSaturation
        )
    }
}
